python/pyarrow/io.pxi

```diff
@@ -101,28 +101,54 @@ cdef class NativeFile:
         """
         raise UnsupportedOperation()
 
+    @property
+    def closed(self):
+        if self.is_readable:
+            return self.input_stream.get().closed()
+        elif self.is_writable:
+            return self.output_stream.get().closed()
+        else:
+            return True
+
     def close(self):
         if not self.closed:
             with nogil:
                 if self.is_readable:
-                    check_status(self.rd_file.get().Close())
+                    check_status(self.input_stream.get().Close())
                 else:
-                    check_status(self.wr_file.get().Close())
-        self.closed = True
+                    check_status(self.output_stream.get().Close())
 
     def flush(self):
         """Flush the buffer stream, if applicable.
 
         No-op to match the IOBase interface."""
         self._assert_open()
 
-    cdef read_handle(self, shared_ptr[RandomAccessFile]* file):
+    cdef set_random_access_file(self, shared_ptr[RandomAccessFile] handle):
+        self.input_stream = <shared_ptr[InputStream]> handle
+        self.random_access = handle
+        self.is_seekable = True
+
+    cdef set_input_stream(self, shared_ptr[InputStream] handle):
+        self.input_stream = handle
+        self.random_access.reset()
+        self.is_seekable = False
+
+    cdef set_output_stream(self, shared_ptr[OutputStream] handle):
+        self.output_stream = handle
+
+    cdef shared_ptr[RandomAccessFile] get_random_access_file(self) except *:
         self._assert_readable()
-        file[0] = <shared_ptr[RandomAccessFile]> self.rd_file
+        self._assert_seekable()
+        return self.random_access
 
-    cdef write_handle(self, shared_ptr[OutputStream]* file):
+    cdef shared_ptr[InputStream] get_input_stream(self) except *:
+        self._assert_readable()
+        return self.input_stream
+
+    cdef shared_ptr[OutputStream] get_output_stream(self) except *:
```

Comment:
Return bare pointers from these functions, will spare a `shared_ptr` copy and fewer `get()` calls in the implementation

Reply:
Ok, will do.

Reply:
Actually, API functions like `get_reader` expect to return a `shared_ptr`, so we can't return a raw pointer here.